projects/btldr/py_client/btldr.py

```diff
@@ -0,0 +1,245 @@
+import math
+import os
+import time
+import logging
+import tqdm
+
+from cantools.database import load_file
+from can import Message as CANMessage
+from can.interface import Bus as CANBus
+
+from .btldr_database import BtldrDatabase, NUM_MESSAGES
+
+
+def _btldr_offset(frame_id):
+    return frame_id % NUM_MESSAGES
+
+
+class BtldrManager:
+    def __init__(self):
+        self.db = BtldrDatabase()
+        logging.info("CAN DBC Initialized")
+
+        # Must be initialized separately
+        self.canbus = None
+
+    def ping(self, ecu_id: int, timeout: int):
+        start = time.time_ns()
+
+        self._send_query(ecu_id)
+        maybe_response = self._receive_query_response(ecu_id, timeout)
+
+        if maybe_response != None:
+            ping_resp = maybe_response
+
+            end = time.time_ns()
+            ping_resp["elapsed_time_ns"] = end - start
+
+            return ping_resp
+        else:
+            return None
+
+    def ping_all(self):
+        raise Exception("Not implemented")
+
+    def flash(self, ecu_id: int, file: str, timeout: int):
+        start = time.time_ns()
+
+        # Reset device into the bootloader
+        self.software_reset(ecu_id, request_update=True)
+
+        time.sleep(1)
+
+        # Query to make sure device is in bootloader
+        ping_resp = self.ping(ecu_id, timeout)
+
+        if not ping_resp:
+            raise Exception("Failed to ping device")
+
+        if ping_resp["current_image"] != "Updater":
+            raise Exception("Failed to place device in updater")
+
+        image_size_bytes = os.path.getsize(file)
+
+        # Image size in 8-byte groups (octets)
+        image_size_octets = math.ceil(float(image_size_bytes) / 8.0)
+
+        # Make image size a multiple of 8
+        image_size_bytes = 8 * image_size_octets
+        self._send_request(ecu_id, "upload", image_size_bytes)
+
+        request_response = self._receive_request_response(ecu_id, timeout)
+
+        if request_response:
+            if request_response["error_code"] != "OK":
```

Comment:
Are there other specific values in request_response["error_code"] that could be useful to include in the raised exception to help while debugging?

Reply:
Nope, it's just OK or ERROR I think.